Jenkinsfile

```diff
@@ -75,7 +101,163 @@ void createPipelineTriggers() {
+ *
+ * @param repo_owner    The owner of the repo.
+ * @param repo_name     The name of the repo that contains the pull request.
+ * @param pr_num        The pull request number.
+ *
+ * @return              [url, commit_id]
+ */
+def getUrlAndCommitFromPr(repo_owner, repo_name, pr_num) {
+    /* Queries depend on the github API v4. To explore the API use:
+       https://developer.github.com/v4/explorer/
+     */
+    /* Escaping in groovy with jenkins is equivalent to the shit from the
+     * illegitimate child of a pile of elephant puke and the still living
+     * cancer from Mussolini. It is horrible...
+     * After about 161 attempts I will say:
+     * - One must use 'single quotes' for credentials as the TOKEN should be
+     * evaluated in the NODE
+     * - Escaping is different from the sh script and a variable
+     * - If one wants to use (parentheses) with sh, one must escape with \\(\\),
+     * this does not apply in variables
+     * - Variables require the expected \ escape characters
+     * - Be careful with what is printed in the output as it doesn't reflect
+     * what is actually called all the time, sometimes an echo will not show
+     * quotes even though they are there...
+     * - It seems like echos show up in label if a url is included, if not
+     * then the label is Print Message and echo data must be expanded
+     * - More info is available: https://gist.github.com/Faheetah/e11bd0315c34ed32e681616e41279ef4
+     *
+     * I wouldn't wish this kind of work on my worst enemy... which right now
+     * seems to be me.
+     */
+    def query = "-X POST -d \"{\\\"query\\\": \\\"query "
+    query = "${query}{repository(name: \\\\\\\"${repo_name}\\\\\\\","
+    query = "${query}owner: \\\\\\\"${repo_owner}\\\\\\\") "
+    query = "${query}{pullRequest(number: ${pr_num}) "
+    query = "${query}{ headRefOid headRepository { url }}}}\\\"}\" "
+    query = "${query}https://api.github.com/graphql"
+    def res = ""
+    withCredentials([string(credentialsId: 'mrkevinweiss_github_api_token',
+            variable: 'TOKEN')]) {
+        res = sh script: '''
+        curl -H "Authorization: token $TOKEN" ''' + query,
+        label: "Fetching commit id and repo", returnStdout: true
+    }
+    def json_obj = readJSON text: res
+    return [json_obj.data.repository.pullRequest.headRepository.url,
+            json_obj.data.repository.pullRequest.headRefOid]
+}
+
+/* Gets the SHA1 commit id and url of the first subrepo in the repo.
+ *
+ * @param repo_owner    The owner of the repo.
+ * @param repo_name     The name of the repo that contains the subrepo.
+ *
+ * @return              [url, commit_id]
+ */
+def getUrlAndCommitOfSubmodule(repo_owner, repo_name) {
+    def query = "-X POST -d \"{\\\"query\\\": \\\"query "
+    query = "${query}{repository(name: \\\\\\\"${repo_name}\\\\\\\","
+    query = "${query}owner: \\\\\\\"${repo_owner}\\\\\\\") "
+    query = "${query}{ submodules(first: 1) "
```

Comment:
Rather than getting the first submodule in the repo, is it possible to get the repo by name instead?

Reply:
It seems not